ouroboros-consensus/src/unstable-consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/Adversarial.hs

```diff
@@ -576,15 +786,42 @@ withinYS (Delta d) !mbYS !(RI.Race (C.SomeWindow Proxy win)) = case mbYS of
 --
 -- The count will be strictly smaller than the number of active slots in the given 'ChainSchema'.
 --
--- REVIEW: why do we not allow forking off the block number 1?
-genPrefixBlockCount :: R.RandomGen g => g -> ChainSchema base hon -> C.Var hon 'ActiveSlotE
-genPrefixBlockCount g schedH =
-    if C.toVar numChoices < 2 then C.Count 0 {- can always pick genesis -} else do
-        C.toVar $ R.runSTGen_ g $ C.uniformIndex numChoices
+-- The result is guaranteed to leave more than k active slots after the
+-- intersection in the honest and the adversarial chains.
+-- See Note [Minimum schema length] in "Test.Ouroboros.Consensus.ChainGenerator.Honest"
+-- for the rationale of the precondition.
+--
+-- PRECONDITION: @schemaSize schedH >= 2*s + d + 1@
+genPrefixBlockCount :: R.RandomGen g => HonestRecipe -> g -> ChainSchema base hon -> C.Var hon 'ActiveSlotE
+genPrefixBlockCount (HonestRecipe (Kcp k) (Scg s) (Delta d) _len) g schedH
+    | C.lengthV v < C.Count (2*s + d + 1) =
+        error "size of schema is smaller than 2*s + d + 1"
+    | numBlocks == C.Count 0 =
+          -- Pick genesis if there are no blocks
+          0
+    | otherwise =
```

Comment:
How about the following?

```
    | otherwise =
        -- @uniformIndex n@ yields a value in @[0..n-1]@, we add 1 to the
        -- argument to account for the possibility of intersecting at the
        -- genesis block
        C.toVar $ R.runSTGen_ g $ C.uniformIndex (phase2 C.+ 1)
  where
    ChainSchema winH vH = schedH

    -- Ultimately, the intersection must have >=k+1 honest asais and >=k+1
    -- adversarial asais.
    --
    -- Our crude approach is as follows; refer to the Note [Minimum schema
    -- length] in "Test.Ouroboros.Consensus.ChainGenerator.Honest".
    --
    -- Let phase1 be the set of honest active slots with at least s+d+k slots
    -- after them. Since we already require 1<k, the first sai can be active.
    -- But those first s sais must also contain the k+1st honest asai, while
    -- only k are guaranteed. Therefore we also drop the youngest member of
    -- phase1, which then serves as the missing +1 for the remaining members.

    phase1 =
        BV.countActivesInV S.notInverted
      $ C.sliceV
          (C.truncateWin winH $ C.windowSize winH C.- (s + k + d))
          vH

    phase2 = phase1 C.- 1
```

It's awkward that we then add `1` back to `phase2` :D, but the comment already explains that.

Reply:
Adopted with edits in 0a406d4.